xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-filters/xwiki-platform-notifications-filters-api/src/main/java/org/xwiki/notifications/filters/internal/DeletedDocumentCleanUpFilterProcessingQueue.java

```diff
@@ -0,0 +1,173 @@
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.notifications.filters.internal;
+
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.component.manager.ComponentLifecycleException;
+import org.xwiki.component.manager.ComponentManager;
+import org.xwiki.component.phase.Disposable;
+import org.xwiki.component.phase.Initializable;
+import org.xwiki.component.phase.InitializationException;
+import org.xwiki.context.concurrent.ExecutionContextRunnable;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.EntityReferenceSerializer;
+import org.xwiki.notifications.NotificationException;
+import org.xwiki.notifications.filters.NotificationFilterPreference;
+import org.xwiki.notifications.filters.NotificationFilterPreferenceManager;
+import org.xwiki.notifications.filters.event.CleaningFilterEvent;
+import org.xwiki.observation.ObservationManager;
+
+/**
+ * Component in charge of processing tasks for cleaning up filters whenever a document is deleted.
+ * This component creates its own low priority thread for this task.
+ * Implementation inspired from {@code PrefilteringLiveNotificationEmailManager}.
+ *
+ * @version $Id$
+ * @since 15.10.2
+ * @since 16.0.0RC1
+ */
+@Component(roles = DeletedDocumentCleanUpFilterProcessingQueue.class)
+@Singleton
+public class DeletedDocumentCleanUpFilterProcessingQueue implements Initializable, Disposable
+{
+    private static class CleanUpFilterData
+    {
+        private final DocumentReference user;
+        private final DocumentReference deletedDocument;
+
+        CleanUpFilterData(DocumentReference user, DocumentReference deletedDocument)
+        {
+            this.user = user;
+            this.deletedDocument = deletedDocument;
+        }
+    }
+
+    private static final CleanUpFilterData STOP_DATA = new CleanUpFilterData(null, null);
+
+    @Inject
+    @Named("context")
+    private ComponentManager componentManager;
+
+    @Inject
+    private EntityReferenceSerializer<String> entityReferenceSerializer;
+
+    @Inject
+    private ObservationManager observationManager;
+
+    @Inject
+    private NotificationFilterPreferenceManager notificationFilterPreferenceManager;
+
+    @Inject
+    private Logger logger;
+
+    private final BlockingQueue<CleanUpFilterData> cleanupQueue = new LinkedBlockingQueue<>();
+
+    private boolean disposed;
+
+    @Override
+    public void dispose() throws ComponentLifecycleException
+    {
+        this.disposed = true;
+        this.cleanupQueue.clear();
```

Comment:
Doesn't this mean that some filters which should be deleted will never be ?

Reply:
That's correct, now not sure how to handle this without redoing what we have for TaskManager.

Reply:
Why not use TaskManager ?

Reply:
Could worth putting a comment about those design choices.

Reply:
I also need a ticket for the actual improvment of TaskManager as IMO it would be useful (would allow also to get rid of the dedicated mechanism for live notification emails: we could use it too, and we wouldn't need a dedicated mechanism for missing emails). 
I guess whenever we do that we should again perform a cleanup of possibly remaining filters though, since as you mentioned we could have some never clean up.

Reply:
>  and we wouldn't need a dedicated mechanism for missing emails

I don't think it makes that much sense as this "dedicated mechanism" is not that dedicated and is relying on the same boolean used by the non-live emails notifications (hourly, daily, etc.) stored in the events core to know if a notification has been sent already or not.

Reply:
I opened https://jira.xwiki.org/browse/XWIKI-21692 so we could move that discussion there